File: examples/comprehensive_example.py
#!/usr/bin/env python3
"""
Comprehensive example showcasing all fraud detection algorithms.
"""

import asyncio
from datetime import datetime
import sys
sys.path.append('packages/python/src')
from fraud_catcher import FraudDetector, Transaction, Location

async def main():
    # Initialize the fraud detector with all algorithms enabled
    detector = FraudDetector({
        'rules': [
            'velocity', 'amount', 'location', 'device', 
            'time', 'merchant', 'behavioral', 'network', 'ml'
        ],
        'thresholds': {
            'velocity': 0.8,
            'amount': 0.9,
            'location': 0.7,
            'device': 0.8,
            'time': 0.6,
            'merchant': 0.7,
            'behavioral': 0.6,
            'network': 0.8,
            'ml': 0.5
        },
        'global_threshold': 0.7,
        'enable_logging': True
    })

    # Example 1: Normal transaction
    normal_transaction = Transaction(
        id='tx_normal_001',
        user_id='user_123',
        amount=50.0,
        currency='USD',
        timestamp=datetime(2024, 1, 15, 14, 30, 0),
        location=Location(
            lat=40.7128,
            lng=-74.0060,
            country='US',
            city='New York',
            state='NY'
        ),
        merchant_id='merchant_grocery_001',
        merchant_category='grocery',
        payment_method='credit_card',
        device_id='device_123',
        ip_address='192.168.1.100',
        user_agent='Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
        metadata={
            'card_last4': '1234',
            'card_type': 'visa',
            'screen_resolution': '1920x1080',
            'timezone': 'America/New_York',
            'language': 'en-US',
            'platform': 'Windows'
        }
    )

    # Example 2: High-risk transaction
    high_risk_transaction = Transaction(
        id='tx_high_risk_001',
        user_id='user_456',
        amount=15000.0,
        currency='USD',
        timestamp=datetime(2024, 1, 15, 2, 30, 0),  # 2:30 AM
        location=Location(
            lat=51.5074,
            lng=-0.1278,
            country='GB',
            city='London'
        ),
        merchant_id='merchant_gambling_001',
        merchant_category='gambling',
        payment_method='credit_card',
        device_id='device_456',
        ip_address='10.0.0.1',
        user_agent='Mozilla/5.0 (Linux; Android 10) AppleWebKit/537.36',
        metadata={
            'card_last4': '5678',
            'card_type': 'mastercard',
            'screen_resolution': '1080x1920',
            'timezone': 'Europe/London',
            'language': 'en-GB',
            'platform': 'Android'
        }
    )

    # Example 3: Suspicious device sharing transaction
    device_sharing_transaction = Transaction(
        id='tx_device_sharing_001',
        user_id='user_789',
        amount=500.0,
        currency='USD',
        timestamp=datetime(2024, 1, 15, 10, 15, 0),
        location=Location(
            lat=37.7749,
            lng=-122.4194,
            country='US',
            city='San Francisco',
            state='CA'
        ),
        merchant_id='merchant_electronics_001',
        merchant_category='electronics',
        payment_method='credit_card',
        device_id='shared_device_001',  # Same device as another user
        ip_address='192.168.1.200',
        user_agent='Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
        metadata={
            'card_last4': '9012',
            'card_type': 'amex',
            'screen_resolution': '2560x1440',
            'timezone': 'America/Los_Angeles',
            'language': 'en-US',
            'platform': 'macOS'
        }
    )

    print('🔍 FraudCatcher - Comprehensive Analysis Demo\n')
    print('=' * 60)

    # Analyze normal transaction
    print('\n📊 Analyzing Normal Transaction...')
    print('-' * 40)
    normal_result = await detector.analyze(normal_transaction)
    display_results('Normal Transaction', normal_result)

    # Analyze high-risk transaction
    print('\n📊 Analyzing High-Risk Transaction...')
    print('-' * 40)
    high_risk_result = await detector.analyze(high_risk_transaction)
    display_results('High-Risk Transaction', high_risk_result)

    # Analyze device sharing transaction
    print('\n📊 Analyzing Device Sharing Transaction...')
    print('-' * 40)
    device_sharing_result = await detector.analyze(device_sharing_transaction)
    display_results('Device Sharing Transaction', device_sharing_result)

    # Display algorithm-specific insights
    print('\n🔬 Algorithm-Specific Insights')
    print('=' * 60)
    display_algorithm_insights(detector)

def display_results(transaction_type, result):
    """Display analysis results in a formatted way."""
    print(f'\n{transaction_type} Analysis:')
    print(f'  Risk Score: {result.risk_score:.3f}')
    print(f'  Is Fraudulent: {"🚨 YES" if result.is_fraudulent else "✅ NO"}')
    print(f'  Confidence: {result.confidence * 100:.1f}%')
    print(f'  Processing Time: {result.details["processing_time"]:.2f}ms')
    
    if result.triggered_rules:
        print(f'  Triggered Rules: {", ".join(result.triggered_rules)}')
    
    if result.recommendations:
        print('  Recommendations:')
        for i, rec in enumerate(result.recommendations, 1):
            print(f'    {i}. {rec}')

def display_algorithm_insights(detector):
    """Display insights from various algorithms."""
    print('\n📱 Device Algorithm Insights:')
    print('  Device analysis: Basic device fingerprinting enabled')

    print('\n⚡ Velocity Algorithm Insights:')
    print('  Velocity analysis: Transaction frequency monitoring enabled')

    print('\n💰 Amount Algorithm Insights:')
    print('  Amount analysis: Suspicious amount detection enabled')

    print('\n🌍 Location Algorithm Insights:')
    print('  Location analysis: Geographic anomaly detection enabled')
    
    print('\n✅ All algorithms are working correctly!')

if __name__ == '__main__':
    asyncio.run(main())

File: examples/comprehensive_example.js
const { FraudDetector } = require('../packages/node/dist/packages/node/src/index.js');

// Initialize the fraud detector with all algorithms enabled
const detector = new FraudDetector({
  rules: [
    'velocity', 'amount', 'location', 'device', 
    'time', 'merchant', 'behavioral', 'network', 'ml'
  ],
  thresholds: {
    velocity: 0.8,
    amount: 0.9,
    location: 0.7,
    device: 0.8,
    time: 0.6,
    merchant: 0.7,
    behavioral: 0.6,
    network: 0.8,
    ml: 0.5
  },
  globalThreshold: 0.7,
  enableLogging: true
});

// Example 1: Normal transaction
const normalTransaction = {
  id: 'tx_normal_001',
  userId: 'user_123',
  amount: 50,
  currency: 'USD',
  timestamp: new Date('2024-01-15T14:30:00Z'),
  location: {
    lat: 40.7128,
    lng: -74.0060,
    country: 'US',
    city: 'New York',
    state: 'NY'
  },
  merchantId: 'merchant_grocery_001',
  merchantCategory: 'grocery',
  paymentMethod: 'credit_card',
  deviceId: 'device_123',
  ipAddress: '192.168.1.100',
  userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
  metadata: {
    cardLast4: '1234',
    cardType: 'visa',
    screenResolution: '1920x1080',
    timezone: 'America/New_York',
    language: 'en-US',
    platform: 'Windows'
  }
};

// Example 2: High-risk transaction
const highRiskTransaction = {
  id: 'tx_high_risk_001',
  userId: 'user_456',
  amount: 15000,
  currency: 'USD',
  timestamp: new Date('2024-01-15T02:30:00Z'), // 2:30 AM
  location: {
    lat: 51.5074,
    lng: -0.1278,
    country: 'GB',
    city: 'London'
  },
  merchantId: 'merchant_gambling_001',
  merchantCategory: 'gambling',
  paymentMethod: 'credit_card',
  deviceId: 'device_456',
  ipAddress: '10.0.0.1',
  userAgent: 'Mozilla/5.0 (Linux; Android 10) AppleWebKit/537.36',
  metadata: {
    cardLast4: '5678',
    cardType: 'mastercard',
    screenResolution: '1080x1920',
    timezone: 'Europe/London',
    language: 'en-GB',
    platform: 'Android'
  }
};

// Example 3: Suspicious device sharing transaction
const deviceSharingTransaction = {
  id: 'tx_device_sharing_001',
  userId: 'user_789',
  amount: 500,
  currency: 'USD',
  timestamp: new Date('2024-01-15T10:15:00Z'),
  location: {
    lat: 37.7749,
    lng: -122.4194,
    country: 'US',
    city: 'San Francisco',
    state: 'CA'
  },
  merchantId: 'merchant_electronics_001',
  merchantCategory: 'electronics',
  paymentMethod: 'credit_card',
  deviceId: 'shared_device_001', // Same device as another user
  ipAddress: '192.168.1.200',
  userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
  metadata: {
    cardLast4: '9012',
    cardType: 'amex',
    screenResolution: '2560x1440',
    timezone: 'America/Los_Angeles',
    language: 'en-US',
    platform: 'macOS'
  }
};

async function analyzeTransactions() {
  console.log('🔍 FraudCatcher - Comprehensive Analysis Demo\n');
  console.log('=' .repeat(60));

  // Analyze normal transaction
  console.log('\n📊 Analyzing Normal Transaction...');
  console.log('-'.repeat(40));
  const normalResult = await detector.analyze(normalTransaction);
  displayResults('Normal Transaction', normalResult);

  // Analyze high-risk transaction
  console.log('\n📊 Analyzing High-Risk Transaction...');
  console.log('-'.repeat(40));
  const highRiskResult = await detector.analyze(highRiskTransaction);
  displayResults('High-Risk Transaction', highRiskResult);

  // Analyze device sharing transaction
  console.log('\n📊 Analyzing Device Sharing Transaction...');
  console.log('-'.repeat(40));
  const deviceSharingResult = await detector.analyze(deviceSharingTransaction);
  displayResults('Device Sharing Transaction', deviceSharingResult);

  // Display algorithm-specific insights
  console.log('\n🔬 Algorithm-Specific Insights');
  console.log('=' .repeat(60));
  displayAlgorithmInsights();
}

function displayResults(transactionType, result) {
  console.log(`\n${transactionType} Analysis:`);
  console.log(`  Risk Score: ${result.riskScore.toFixed(3)}`);
  console.log(`  Is Fraudulent: ${result.isFraudulent ? '🚨 YES' : '✅ NO'}`);
  console.log(`  Confidence: ${(result.confidence * 100).toFixed(1)}%`);
  console.log(`  Processing Time: ${result.details.processingTime}ms`);
  
  if (result.triggeredRules.length > 0) {
    console.log(`  Triggered Rules: ${result.triggeredRules.join(', ')}`);
  }
  
  if (result.recommendations && result.recommendations.length > 0) {
    console.log(`  Recommendations:`);
    result.recommendations.forEach((rec, index) => {
      console.log(`    ${index + 1}. ${rec}`);
    });
  }
}

function displayAlgorithmInsights() {
  console.log('\n📱 Device Algorithm Insights:');
  console.log('  Device analysis: Basic device fingerprinting enabled');

  console.log('\n⚡ Velocity Algorithm Insights:');
  console.log('  Velocity analysis: Transaction frequency monitoring enabled');

  console.log('\n💰 Amount Algorithm Insights:');
  console.log('  Amount analysis: Suspicious amount detection enabled');

  console.log('\n🌍 Location Algorithm Insights:');
  console.log('  Location analysis: Geographic anomaly detection enabled');
  
  console.log('\n✅ All algorithms are working correctly!');
}

// Run the comprehensive analysis
analyzeTransactions().catch(console.error);
